path/walker.go

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package path
+
+// Type Visitor is a function which performs an action on the provided path.
+// The return value is a slice of subdirectories to visit, a function to call to "close"
+// the visited directory on successful visitation or an error.
+type Visitor func(string) ([]string, func() error, error)
```

Comment:
This interface feels a little weird to me—perhaps in part because I can't see a clear reason why the caller of the visitor should need to defer the close. Once the visitor has determined the names of the subdirectories, does the caller's ability to open those files depend on the directory itself still being "open"? I assume there's a motivating example, but I wasn't able to infer it from the usage.

Reply:
The motivation was to provide clients the ability to do pre- and post- order visitation with a non-recursive traversal, e.g.
```
visit(...) {
  log.Println("Open:", dir)
  ...
  return children, func() { log.Println("Close:", dir), nil
}
```

Reply:
Is there some reason _not_ to do a recursive traversal?

I think my main concern here is that it's hard to understand how to write the callback for `Walk`, and in particular how to write the callback that the callback has to return. If it's important to consider traversal order, maybe this should be more than just a plain function, e.g.,

```go
type Visitor interface {
   Enter(path string) ([]string, error) // returns names of children to visit
   Leave(path string, kids []string) error // called after all children are processed
}
```

A preorder traversal can implement everything in `Enter` and have `Leave` be a no-op. A postorder traversal can just return the children and then do its cleanup in `Leave`. For the trivial case, you can shim in a function:

```go
type PreVisitor func(string) ([]string, error)

func (p PreVisitor) Enter(path string) ([]string, error) { return p(path) }
func (PreVisitor) Leave(string, []string) error { return nil }
```

I think this would eliminate the need for a two-way callback plumbing, and still allow non-recursive traversal with either visitation order. I didn't actually test this, though.

```go
if err := path.Walk(root, path.PreVisitor(doTheThing)); err != nil {
  log.Fatalf("Walk: %v", err)
}
```

Reply:
Mostly just principle on the traversal, the implementation is currently recursive (but I wanted to allow for something different).  The full interface is easier to understand, so I've mostly made that change, thanks!

I think I'd also wanted it to be easier to pass any required "state" around, rather than dictating the interface implementations store that themselves, but :man_shrugging: